### PR TITLE
Gcehealthchecks

### PR DESCRIPTION
**Pull request by ILMTitan**
Add cmdlets for Add, Get, Set, and Remove GceHealthChecks.

### CODE REVIEW

File: Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        $healthChecks.Name | Should Be $healthCheckName
+        $healthChecks.Port | Should Be 50
+        $healthChecks.Host | Should Be "google.com"
+        $healthChecks.RequestPath | Should Be "/some/path"
+        $healthChecks.CheckIntervalSec | Should Be 2
+        $healthChecks.TimeoutSec | Should Be 2
+        $healthChecks.HealthyThreshold | Should Be 3
+        $healthChecks.UnhealthyThreshold | Should Be 3
+    }
+}
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1]: This test calls `Set-GCloudConfig` but does not restore the Cloud configuration at the end.

File: Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            $healthCheck.Host | Should Be "google.com"
+            $healthCheck.RequestPath | Should Be "/some/path"
+            $healthCheck.CheckIntervalSec | Should Be 2
+            $healthCheck.TimeoutSec | Should Be 2
+            $healthCheck.HealthyThreshold | Should Be 3
+            $healthCheck.UnhealthyThreshold | Should Be 3
+
+        }
+
+        It "should use an http object over pipeline" {
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1]: HTTP is an acronym, so all caps.

File: Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+    It "should fail for wrong project" {
+        { Get-GceHealthCheck -Project "asdf" } | Should Throw 403
+    }
+
+    It "should fail for non-existant health check" {
+        { Get-GceHealthCheck $healthCheckName -Http } | Should Throw 404
+    }
+
+    It "should successfully list zero" {
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1]: "list zero" doesn't make sense without any context. How about "retrieve all health checks for a project".

And the first scenario of the test confirms there are 0, then it adds 2, and confirms the count is then 2.

File: Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        $noChecks = Get-GceHealthCheck
+        $noChecks.Count | Should Be 0
+    }
+
+    Context "with checks" {
+        BeforeAll {
+            Add-GceHealthCheck $healthCheckName
+            Add-GceHealthCheck $healthCheckName2
+            Add-GceHealthCheck $healthCheckName -Https
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1]: Maybe comment here? Does this add a third health check, because the name isn't not unique, it is name x HTTPS?

File: Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+Describe "Set-GceHealthCheck" {
+    $healthCheckName = "test-set-gcehealthcheck-$r"
+
+    Add-GceHealthCheck $healthCheckName
+    Add-GceHealthCheck $healthCheckName -Https
+
+    AfterAll {
+        Get-GceHealthCheck |
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1]: Can this fit on the same line?

File: Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+            Remove-GceHealthCheck
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/Compute/Compute.GceHealthCheck.Tests.ps1]: Put this on the same line as above?

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+using System.Management.Automation;
+
+namespace Google.PowerShell.ComputeEngine
+{
+    /// <summary>
+    /// <para type="synopsis">
+    /// Adds a Google Compute Engine health check.
+    /// </para>
+    /// <para type="descritpion">
+    /// Adds a Google Compute Engine health check. Use this cmdlet to create both http and https health checks.
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: HTTP and HTTPS. Also, is there a link to the docs you can point to? i.e. something that describes what health checks are used for with examples?

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+        /// <para type="description">
+        /// Human readable description of the health check.
+        /// </para>
+        /// </summary>
+        [Parameter(ParameterSetName = ParameterSetNames.ByValues)]
+        public string Description { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// The value of the host header in the health check request. If left empty
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: "If left empty (default value)" isn't cromulent. Could you rephrase this to be easier to parse?

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+        }
+    }
+
+    /// <summary>
+    /// <para type="synopsis">
+    /// Gets Google Compute Engine health checks.
+    /// </para>
+    /// <para type="description">
+    /// Gets Google Compute Engine health checks.
+    /// This cmdlet can be used to retrieve both http and https health checks.
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: Mention it can get all checks for a project, or a subset based on the name, etc.

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+        /// <para type="description">
+        /// The name of the project to get the health checks of.
+        /// Defaults to the project in the Cloud SDK config.
+        /// </para>
+        /// </summary>
+        [Parameter]
+        [ConfigPropertyName(CloudSdkSettings.CommonProperties.Project)]
+        public string Project { get; set; }
+
+
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: Vertical whitespace.

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+            if (healthChecks.Count == 0)
+            {
+                if (exceptions.Count == 1)
+                {
+                    throw exceptions[0];
+                }
+                else
+                {
+                    throw new AggregateException(
+                        $"The health check named {Name} in project {Project} can not be found", exceptions);
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: Should exceptions be written in the past tense? e.g. "could not be found"?

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+        /// The name of the health check to remove.
+        /// </para>
+        /// </summary>
+        [Parameter(ParameterSetName = ParameterSetNames.ByNameHttp, Mandatory = true, Position = 0)]
+        [Parameter(ParameterSetName = ParameterSetNames.ByNameHttps, Mandatory = true, Position = 0)]
+        public string Name { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// If set, will remove a health check that uses http.
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: Mmm, I love me some HTTP in all caps.

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 13]
```diff
@@ -0,0 +1,545 @@
...
+                    RemoveHttpsHealthCheck(GetProjectNameFromUri(HttpsObject.SelfLink), HttpsObject.Name);
+                    break;
+                default:
+                    throw UnknownParameterSetException;
+            }
+        }
+
+        private void RemoveHttpsHealthCheck(string project, string name)
+        {
+            if (ShouldProcess($"{project}/{name}", "Remove Https Health Check"))
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: One fourth of the way there ;)

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 14]
```diff
@@ -0,0 +1,545 @@
...
+            {
+                HttpsHealthChecksResource.DeleteRequest request = Service.HttpsHealthChecks.Delete(project, name);
+                Operation operation = request.Execute();
+                AddGlobalOperation(project, operation);
+            }
+        }
+
+        private void RemoveHttpHealthCheck(string project, string name)
+        {
+            if (ShouldProcess($"{project}/{name}", "Remove Http Health Check"))
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: When I saw the integration tests, I saw we didn't need to pass `-Confirm` or `-Force`. I assume this is because Pester auto-confirms prompts when we call `ShouldProcess`?

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 15]
```diff
@@ -0,0 +1,545 @@
...
+        }
+    }
+
+    /// <summary>
+    /// <para type="synopsis">
+    /// Sets the data of a Google Compute Engine health check.
+    /// </para>
+    /// <para type="description">
+    /// Sets the data of a Google Compute Engine health check.
+    /// First get the health check object with Get-GceHealthCheck.
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: You can put this on the line above. Otherwise the formatting might get weird. Just write paragraphs in long-form, breaking at 120 columns.

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 16]
```diff
@@ -0,0 +1,545 @@
...
+    /// <summary>
+    /// <para type="synopsis">
+    /// Removes a Google Compute Engine health check.
+    /// </para>
+    /// <para type="description">
+    /// Removes a Google Compute Engine Health check. Use this cmdlet to remove both http and https health
+    /// checks.
+    /// </para>
+    /// </summary>
+    [Cmdlet(VerbsCommon.Remove, "GceHealthCheck", SupportsShouldProcess = true)]
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: Probably good to add an example here. Remove all HealthChecks matching a specific string and passing `-WhatIf`?

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 17]
```diff
@@ -0,0 +1,545 @@
...
+
+    /// <summary>
+    /// <para type="synopsis">
+    /// Gets Google Compute Engine health checks.
+    /// </para>
+    /// <para type="description">
+    /// Gets Google Compute Engine health checks.
+    /// This cmdlet can be used to retrieve both http and https health checks.
+    /// </para>
+    /// </summary>
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: Add a few examples, getting all health checks for a project, just HTTPS with a certain name.

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 18]
```diff
@@ -0,0 +1,545 @@
...
+        protected override void ProcessRecord()
+        {
+            string project;
+            string name;
+            switch (ParameterSetName)
+            {
+                case ParameterSetNames.ByHttpObject:
+                    project = GetProjectNameFromUri(HttpObject.SelfLink);
+                    name = HttpObject.Name;
+                    if (ShouldProcess($"{project}/{name}", "Set Http Health Check"))
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: For other for `Set-` based cmdlets, I don't recall if we call `ShouldProcess`. Should we be more consistent here? Or are we just doing that here because HTTP health checks are kinda important, unlike tweaking metadata for GCS objects.

**Comment by chrsmith** [comment]
No major changes, but I'd like to understand the answer to a few of the questions before you submit.

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs (hunk 14 above)
Comment (ILMTitan) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: It is because the default $ConfirmPreference level is High enough to not cause a popup.

**Comment by ILMTitan** [comment]
PTAL

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 19]
```diff
@@ -12,8 +12,29 @@ namespace Google.PowerShell.ComputeEngine
...
+    /// Adds a Google Compute Engine health check. Use this cmdlet to create both HTTP and HTTPS health checks.
+    /// https://cloud.google.com/compute/docs/load-balancing/health-checks
     /// </para>
+    /// <example>
+    ///   <para> Adds an HTTP health to project "my-project". </para>
+    ///   <para><code> Add-GceHealthCheck "my-health-check" -Project "my-project" -Http </code></para>
+    /// </example>
+    /// <example>
+    ///   <para> Adds an HTTP health to project "my-project". </para>
+    ///   <para><code> Add-GceHealthCheck "my-health-check" -Project "my-project" -Http </code></para>
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: Microsoft's PowerShell examples are prefixed with "`PS C:\> `". This actually helps differentiate which is a command and what is text around it. 

Could you please change this and others to:

///   <para><code>PS C:\> Add-GceHealthCheck "my-health-check" -Project "my-project" -Http </code></para>

**Comment by ILMTitan** [comment]
PTAL

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 20]
```diff
@@ -16,24 +16,20 @@ namespace Google.PowerShell.ComputeEngine
...
-    ///     PS C:&gt; Add-GceHealthCheck "my-health-check" -Http -Description "Description of my health check." `<br/>
-    ///         -HostHeader "mydomain.com" -Port 50 -RequestPath "/some/path" -CheckInterval "0:0:2" `<br/>
-    ///         -Timeout "0:0:2" -HealthyThreshold 3 -UnhealthyThreshold 3 `<br/>
-    ///   </code></para>
+    ///   <code>
+    ///     <para>PS C:&gt; Add-GceHealthCheck "my-health-check" -Http -Description "Description of my health check." `</para>
+    ///     <para>            -HostHeader "mydomain.com" -Port 50 -RequestPath "/some/path" -CheckInterval "0:0:2" `</para>
+    ///     <para>            -Timeout "0:0:2" -HealthyThreshold 3 -UnhealthyThreshold 3 `</para>
+    ///   </code>
+    ///     <para> Adds an HTTP health check with non-default values</para>
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: The convention other examples have been using is to have a description before the code snippet. Also, we have been using `<para><code>` not `<code><para>`. I don't know if that is the same, better, or worse. Just different.

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs [hunk 21]
```diff
@@ -215,18 +211,18 @@ protected override void ProcessRecord()
     /// of a project, or get a health check by name.
     /// </para>
     /// <example>
+    ///   <code> PS C:&gt; Get-GceHealthCheck -Project "my-project" </code>
```
Comment (chrsmith) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: Looks like we'll need to s/C:\/C:&gt;/ for all cmdlets...

**Comment by chrsmith** [comment]
LGTM

File: Google.PowerShell/Compute/GceHealthCheckCmdlets.cs (hunk 20 above)
Comment (ILMTitan) [inline Google.PowerShell/Compute/GceHealthCheckCmdlets.cs]: The examples from the core cmdlets (and the Xml2cmdletDoc documentation) has the example before the description.
